src/nnet3/nnet-cudnn-simple-component.cc

```diff
@@ -0,0 +1,798 @@
+  strides[1] = filt_y_stride;
+  strides[2] = filt_x_stride;
+
+  int32 upscales[kConvolutionDimension_];
+  upscales[0] = upscale_z_dim;
+  upscales[1] = upscale_y_dim;
+  upscales[2] = upscale_x_dim;
+
+  int32 padding[kConvolutionDimension_];
+  padding[0] = pad_z_dim;
+  padding[1] = pad_y_dim;
+  padding[2] = pad_x_dim;
+
+  CUDNN_SAFE_CALL(
+    cudnnSetConvolutionNdDescriptor(conv_desc_,
+                                    kConvolutionDimension_,
+                                    padding,
+                                    strides,
+                                    upscales,
+                                    CUDNN_CROSS_CORRELATION,
+                                    cudnn::GetDataType())
+  );
+
+  CUDNN_SAFE_CALL(cudnnCreateFilterDescriptor(&filter_desc_));
+  CUDNN_SAFE_CALL(
+    cudnnSetFilterNdDescriptor(filter_desc_,
+                               cudnn::GetDataType(),
+                               kConvolutionDimension_ + 2,
+                               filters
+                               )
+  );
+
+  int32 bias_dims[kConvolutionDimension_ + 2] = {
+    1,
+    num_filters_,
+    1,
+    1,
+    1
+  };
+
+  int32 bias_strides[kConvolutionDimension_ + 2] = {
+    num_filters_,
+    1,
+    1,
+    1,
+    1
+  };
+
+  CUDNN_SAFE_CALL(
+    cudnnSetTensorNdDescriptor(bias_desc_,
+                               cudnn::GetDataType(),
+                               kConvolutionDimension_ + 2,
+                               bias_dims,
+                               bias_strides
+                               )
+                  );
+
+
+}
+
+void CuDNN3DConvolutionComponent::InitFromConfig(ConfigLine *cfl) {
+  bool ok = true;
+  int32 input_x_dim = -1, input_y_dim = -1, input_z_dim = -1,
+    input_num_filters = -1,
+    filt_x_dim = -1, filt_y_dim = -1, filt_z_dim = -1,
+    filt_x_stride = -1, filt_y_stride = -1, filt_z_stride = -1,
+    num_filters = -1,
+    upscale_x_dim = -1, upscale_y_dim = -1, upscale_z_dim = -1,
+    pad_x_dim = -1, pad_y_dim = -1, pad_z_dim = -1;
+
+  std::string input_vectorization_order = "zyx";
+  TensorVectorizationType input_vectorization = kZyx;
+  // TODO: Figure our whether we need to allow input_vectorization_order
+  // to be configurable.
+
+  InitLearningRatesFromConfig(cfl);
+  ok = ok && cfl->GetValue("num-filters", &num_filters);
+  ok = ok && cfl->GetValue("input-num-filters", &input_num_filters);
+  ok = ok && cfl->GetValue("input-x-dim", &input_x_dim);
+  ok = ok && cfl->GetValue("input-y-dim", &input_y_dim);
+  ok = ok && cfl->GetValue("input-z-dim", &input_z_dim);
+  ok = ok && cfl->GetValue("filt-x-dim", &filt_x_dim);
+  ok = ok && cfl->GetValue("filt-y-dim", &filt_y_dim);
+  ok = ok && cfl->GetValue("filt-z-dim", &filt_z_dim);
+  ok = ok && cfl->GetValue("filt-x-stride", &filt_x_stride);
+  ok = ok && cfl->GetValue("filt-y-stride", &filt_y_stride);
+  ok = ok && cfl->GetValue("filt-z-stride", &filt_z_stride);
+
+  // upscale_<k>_dim is how many times to
```

Comment:
@galv , is filt-x-stride has the same meaning as filt-x-step in the original Convolution Component ?